### Assignment 

### Question 1

Q1. What is the difference between Ordinal Encoding and Label Encoding? Provide an example of when you
might choose one over the other

__Answer__

The major difference between both is that one is use to code catgorical variables with natural occurring order e.g levels of education __(BSc, MSc, PhD)__ while Label Encoding if used for coding categorical variables that does not follow a natural order e.g color __(red, white, green)__


__Further Explanation__

__Label Encoding__ assigns a unique numerical value to each category of a categorical variable, with no consideration of any order or hierarchy among them. For example, in a dataset with a "color" variable, label encoding would assign the values "red", "green", and "blue" to the values 0, 1, and 2 respectively.

__Ordinal Encoding,__ on the other hand, assigns numerical values to the categories based on their order or hierarchy. For example, if the "size" variable in a dataset has categories "small", "medium", and "large", ordinal encoding might assign the values 0, 1, and 2 respectively.

### Question 2

Q2. Explain how Target Guided Ordinal Encoding works and provide an example of when you might use it in
a machine learning project.

__Answer__

__Target Guided Ordinal Encoding__ is a technique for encoding categorical data by using the target variable. In this technique, the categories are ordered based on the mean of the target variable corresponding to each category. The categories are then assigned an ordinal number based on this ordering1.

__Example__

A dataset containing information about employees, including their city location and salary. The city location variable can be coded using Target Guided Ordinal Encoding by first calculating the mean salary for each city. Then order the  cities based on their mean salary and assign an ordinal number to each city based on this ordering.

Note: mean also can be use directly to represent each city.

### Question 


_Q3. Define covariance and explain why it is important in statistical analysis. How is covariance calculated?

_Answer__

Covariance is a statistical measure that indicates the extent to which two variables are related to each other. Specifically, covariance measures how much two variables vary together or co-vary. A positive covariance indicates that the two variables increase or decrease together, while a negative covariance indicates that as one variable increases, the other decreases. A covariance of zero indicates that there is no linear relationship between the two variables.

__Importance__

Covariance is important in statistical analysis because it helps to identify how two variables are related to each other. It provides information about the strength and direction of the relationship between two variables. Covariance is commonly used in finance to measure the risk and return of a portfolio. In data analysis, covariance is used to identify the strength of the relationship between two variables and to help select variables for regression analysis.

Mathematically, the formula for covariance between two variables X and Y is:

Cov(X, Y) = (Σ[(X - μX) * (Y - μY)]) / (n - 1)

where:

μX is the mean of variable X

μY is the mean of variable Y

n is the number of observations

The result of the formula is a single number that represents the degree to which the two variables covary. A positive value indicates that the two variables have a positive relationship, while a negative value indicates a negative relationship. The magnitude of the covariance is difficult to interpret because it depends on the units of measurement of the variables. To overcome this, correlation, which is a standardized version of covariance, is often used.


### Question 4

Q4. For a dataset with the following categorical variables: Color (red, green, blue), Size (small, medium, large), and Material (wood, metal, plastic), perform label encoding using Python's scikit-learn library.
Show your code and explain the output.

__Answer__

check the code below:

In [37]:
## The categorical data

color = ["red", "green", "blue"]
size = ["small", "medium", "large"]
material = ["wood", "metal", "plastic"]

import pandas as pd

df = pd.DataFrame({'color': color, "size": size, 'material': material})

df

color    size material
0    red   small     wood
1  green  medium    metal
2   blue   large  plastic

In [38]:
### calling in the library for the encoding and loading for label
from sklearn.preprocessing import LabelEncoder
lbl_encoder=LabelEncoder()  ## creating and insance for method

In [44]:
## label encode color
df['color_enc'] = lbl_encoder.fit_transform(df['color'])

## label encode size
df['size_enc'] = lbl_encoder.fit_transform(df['size'])

# label encode material
df['mat_enc'] = lbl_encoder.fit_transform(df['material'])


# printing only the encodings out
print(df[['color_enc','size_enc',  'mat_enc']])

print('''\nLabel encoding involves assigning a unique numerical 
label to each category in the variable. 
The labels are usually assigned in alphabetical 
order or based on the frequency of the categories.''')

   color_enc  size_enc  mat_enc
0          2         2        2
1          1         1        0
2          0         0        1

Label encoding involves assigning a unique numerical 
label to each category in the variable. 
The labels are usually assigned in alphabetical 
order or based on the frequency of the categories.


### Question 5

Q5. Calculate the covariance matrix for the following variables in a dataset: Age, Income, and Education level. Interpret the results.

__Answer__



In [46]:
import numpy as np
import pandas as pd

# Set the random seed for reproducibility
np.random.seed(123)

# Generate random data for Age, Income, and Education
n = 1000
age = np.random.normal(loc=35, scale=10, size=n) ## loc==mean, scale == s.deviation,
income = np.random.normal(loc=50000, scale=10000, size=n)
education = np.random.normal(loc=12, scale=2, size=n)

# Create a Pandas DataFrame with the generated data
data = pd.DataFrame({'Age': age, 'Income': income, 'Education': education})

# Calculate the covariance matrix
cov_matrix = data.cov()

print(cov_matrix)

## note this are hypothetical numbers not to be interpreted as real data

                   Age        Income   Education
Age         100.257827 -3.002349e+03   -0.637299
Income    -3002.349265  9.187010e+07 -252.094324
Education    -0.637299 -2.520943e+02    3.856899


Interpreting the results:

1. The diagonal elements of the covariance matrix show the variance of each variable (e.g., the variance of Age is 100.257827).
2. The off-diagonal elements show the covariance between the corresponding pairs of variables (e.g., the covariance between Age and Income is -3002.349265).
3. A positive covariance value indicates that the two variables tend to increase or decrease together, while a negative covariance indicates that the two variables tend to move in opposite directions.
4. The magnitude of the covariance values does not tell us whether the relationship between the variables is strong or weak. To measure the strength of the relationship, we need to calculate the correlation coefficient.

### Question 6

Q6. You are working on a machine learning project with a dataset containing several categorical  variables, including "Gender" (Male/Female), "Education Level" (High School/Bachelor's/Master's/PhD), and "Employment Status" (Unemployed/Part-Time/Full-Time). Which encoding method would you use for
each variable, and why?


__Answer__

__Gender (binary categorical variable):__

One-Hot Encoding  ==> I will use this becuase this feature has two categories and one hot encoding which is binary 0 and 1 is best suited for this. Since it will indicate properly if the respondant is either male or female with proper separtation

__Education Level (ordinal categorical variable):__

Ordinal Encoding == > the level of the categories in education follow natural order of significant and moreso education is stepwise. each level will be mapped to numbers base on order. For example, we can assign the value 1 to High School, 2 to Bachelor's, 3 to Master's, and 4 to PhD. Ordinal encoding preserves the order of the categories and is a good choice when there is a natural ordering among the categories.

__Employment Status (nominal categorical variable):__

Label Encoding ==> since no natural order to the level and number of categories is small. Label encoding involves assigning a unique numerical label to each category in the variable. The labels are usually assigned in alphabetical order or based on the frequency of the categories.


### Question 7

Q7. You are analyzing a dataset with two continuous variables, "Temperature" and "Humidity", and two categorical variables, "Weather Condition" (Sunny/Cloudy/Rainy) and "Wind Direction" (North/South/East/West). Calculate the covariance between each pair of variables and interpret the results.
__Answer__


In [48]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder

# Create a sample dataset
data = {
    'Temperature': [25.0, 27.0, 22.0, 20.0, 24.0],
    'Humidity': [40.0, 45.0, 50.0, 55.0, 60.0],
    'Weather Condition': ['Sunny', 'Cloudy', 'Rainy', 'Sunny', 'Cloudy'],
    'Wind Direction': ['North', 'South', 'East', 'West', 'South']
}

df = pd.DataFrame(data)

# Label encode categorical variables
le = LabelEncoder()
df['Weather Condition'] = le.fit_transform(df['Weather Condition'])
df['Wind Direction'] = le.fit_transform(df['Wind Direction'])

# Calculate the covariance matrix
cov_matrix = df.cov()

print(cov_matrix)


                   Temperature  Humidity  Weather Condition  Wind Direction
Temperature               7.30    -11.25               -1.5           -0.45
Humidity                -11.25     62.50               -2.5            3.75
Weather Condition        -1.50     -2.50                1.0            0.00
Wind Direction           -0.45      3.75                0.0            1.30


__Explanation__

I label encode the categorical variable before running the covariance analysis.

Covariance between temperature and humidity is negative, which means that as temperature increases, humidity tends to decrease. The covariance between weather condition and temperature/humidity is small, indicating that there may not be a strong relationship between them. The covariance between wind direction and temperature/humidity is also small, indicating that wind direction may not have a strong influence on temperature and humidity.